In [1]:
import tensorflow as tf
print("TensorFlow Version: {}".format(tf.__version__))

TensorFlow Version: 2.3.0


In [2]:
# load dataset
mnist = tf.keras.datasets.mnist

In [3]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

The idea below is to convert from integers to floating numbers.

In [4]:
x_train, x_test = x_train/255.0, x_test/255.0

Build a model using keras Sequential API and stack layers.

In [5]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

The model returns a vector of logits and log-odds scores.

In [6]:
predictions = model(x_train[:1]).numpy()
predictions

array([[-0.5035181 ,  0.10101207,  0.6072018 , -0.0099389 , -0.22559053,
        -0.15243143,  0.18064716, -0.32102472, -0.19315434,  0.11745246]],
      dtype=float32)

Softmax function converts the logits to probabilities for each class.

In [7]:
tf.nn.softmax(predictions).numpy()

array([[0.06004894, 0.1099131 , 0.18234123, 0.09837032, 0.07928805,
        0.08530615, 0.119024  , 0.07207111, 0.08190202, 0.11173505]],
      dtype=float32)

Defined loss function takes a vector of logits and a True index and returns a scalar loss for each example. <br><br>
The loss is equal to the negative log probability of the true class. The loss is zero if the model is sure of the correct class.

In [8]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [9]:
loss_fn(y_train[:1], predictions).numpy()

2.4615088

Configure and compile the model.

In [10]:
model.compile(optimizer="adam", loss=loss_fn, metrics=["accuracy"])

Fit the model to adjust model parameters and minimize the loss

In [11]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2981 - accuracy: 0.9134
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1452 - accuracy: 0.9576
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1080 - accuracy: 0.9665
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0884 - accuracy: 0.9717
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0750 - accuracy: 0.9762


Check performance on a validation set or a test set

In [12]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.0793 - accuracy: 0.9750


[0.079252228140831, 0.9750000238418579]

Model trained to 98% accuracy on the dataset.<br><br>
Credits: https://www.tensorflow.org/tutorials/quickstart/beginner